src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SemanticFacts/CSharpSemanticFacts.cs

```diff
@@ -83,6 +83,15 @@ public bool CanReplaceWithRValue(SemanticModel semanticModel, [NotNullWhen(true)
                 // If we hit an executable statement syntax and didn't find anything yet, we can just stop now -- anything higher would be a member declaration which won't be defined by something inside a statement.
                 if (CSharpSyntaxFacts.Instance.IsExecutableStatement(ancestor))
                     return null;
+
+                switch (ancestor.Kind())
+                {
+                    case SyntaxKind.IfDirectiveTrivia:
+                    case SyntaxKind.ElifDirectiveTrivia:
+                    case SyntaxKind.DefineDirectiveTrivia:
+                    case SyntaxKind.UndefDirectiveTrivia:
+                        return semanticModel.Compilation.CreatePreprocessingSymbol(token.ValueText);
```

Comment:
what's this for?

Reply:
`SymbolFinder.GetSemanticInfoAtPositionAsync` uses `SemanticModel.GetSemanticInfo` to determine the semantic info for the symbol at the specified position. This is useful for the FAR feature, as it would not retrieve semantic information otherwise. This specific implementation utilizes the already existing parent traversal to determine whether we are in a preprocessor directive that accepts pp symbols.

Reply:
`GetPreprocessingSymbolInfo` retrieves the preprocessing symbol's information, including whether it is defined in the current position of the token. This imposes some computation overhead for a result that we are going to discard, since we only care about the symbol appearing as a result.

However I still do not like the fact that I'm doing this breaking change in `GetDeclaredSymbol`. In none of the above syntax kinds is a pp symbol actually being declared, to my understanding pp symbols are never truly declared. Thus, the ref finder should be able to look for that pp symbol from the given token without depending on this method.

Reply:
i still don't get it.

> This imposes some computation overhead for a result that we are going to discard, since we only care about the symbol appearing as a result.

I don't get why i would care about that.  If we have a semantic quesiton, we make a semantic request.  If htat has cpu cost, so be it.

> However I still do not like the fact that I'm doing this breaking change in GetDeclaredSymbol.

What breaking change?

>  In none of the above syntax kinds is a pp symbol actually being declared,

Then why are you updating this method?  I don't get it.

Reply:
> I don't get why i would care about that. If we have a semantic quesiton, we make a semantic request. If htat has cpu cost, so be it.

We still discard the result so it's unnecessary to do all that work. If we truly had to use that path, I would still be in favor of separating the work for finding whether the symbol is currently defined in the span that it appears, and the symbol itself that is represented by the syntax node.

> What breaking change?

By supporting these specific syntax kinds, we are returning a declared symbol that was previously not returned. To me that looks like a breaking change, since we are adjusting the behavior on a specific input in a publicly exposed method. Users could break by not expecting a preprocessing symbol and encounter issues after upgrading their dependencies.

> Then why are you updating this method? I don't get it.

I'm considering adjusting this change to target the support for finding preprocessing symbols' references without needing to define them as being "declared" in the places they are being used, including `if` directives. The goal is to not depend solely on `GetDeclaredSymbol` and specifically handle preprocessing symbols for FAR.

Reply:
> We still discard the result so it's unnecessary to do all that work

* why does this matter * ?

We make tons of calls to the compiler to get an answer that we only use partial data from (like information about nodes, symbols, operations, etc.).  That's just how query APIs work.  I haven't seen a single piece of information indicating that this is a problem here to get this information for teh *exceptionally* rare case that someone would be asking about PP data.

Reply:
> Users could break by not expecting a preprocessing symbol and encounter issues after upgrading their dependencies.

This is how roslyn has always worked.  We return new information on nodes we didn't provide information on in the past.

>  The goal is to not depend solely on GetDeclaredSymbol and specifically handle preprocessing symbols for FAR.

That would be wonderful :) I think that was something i asked for very early on.  It's never been clear to me why such a simple FAR change has involved so many changes to so many pieces that don't seem involved :-/